java/org/apache/catalina/Realm.java

```diff
@@ -116,6 +118,17 @@ public Principal authenticate(String username, String digest,
     public Principal authenticate(GSSContext gssContext, boolean storeCreds);
 
 
+    /**
+     * Try to authenticate using a {@link GSSName}
+     *
+     * @param gssName The {@link GSSName} of the principal to look up
+     * @param gssCredential The {@link GSSCredential} of the principal, may be
+     *                      {@code null}
+     * @return the associated principal, or {@code null} if there is none
+     */
+    public Principal authenticate(GSSName gssName, GSSCredential gssCredential);
+
```

Comment:
Where does this method get used? What is its purpose?

If this PR is applied, I think this should have a default implementation that returns null.

Reply:
Please look into `org.apache.catalina.realm.RealmBase.authenticate(GSSName, GSSCredential)` it will properly handle the `null` case. Those who extend `RealmBase` won't see a change.

The purpose has been described in detail here: https://bz.apache.org/bugzilla/show_bug.cgi?id=63681#c0

Reply:
And those who don't extend RealmBase will see breakage. Please add the default method.

Reply:
That is true, but that would ultimately mean that it would be available in Tomcat 9+ only. SPNEGO support has been introduced with BZ 48685 which was fist implemented in 577c4e6194b71a63be70725bb743c94c7cb082cc and was included in these tags:

```
$ git tag --contains 577c4e6194b71a63be70725bb743c94c7cb082cc
7.0.12
7.0.13
7.0.14
7.0.15
7.0.16
7.0.17
7.0.18
7.0.20
9.0.0
9.0.0-M1
9.0.0-M10
9.0.0-M11
9.0.0-M12
9.0.0-M13
9.0.0-M14
9.0.0-M15
9.0.0-M16
9.0.0-M17
9.0.0-M18
9.0.0-M19
9.0.0-M2
9.0.0-M20
9.0.0-M21
```

which basically means that the API has already been broken. I wouldn't mind to provide this to 8.5.x and 9.0.x only. Since 8.5.x and 9.0.x are interchangeable expect Servlet API it would be whise to have it either in both or in none.

Reply:
@markt-asf So you think that even with the fact above this cannot be reasonably added to 8.5+ w/o breaking the API?!

Reply:
Yes.

Reply:
OK, I see two alternatives here:

* Drop the change, move it to a change until Tomcat 10 emerges
* Introduce a `GSSRealm` interface which extends `Realm` and `ReamBase` will implement `GSSRealm`. All Tomcat-provided authenticators will still use `Realm` and custom authenticators can decide. I will also mark the interface as deprecated for removal in Tomcat 10.

How does that sound? Which of those would you prefer?

Reply:
Or...
Add method with default implementation to `Realm` in 9.0.x and add `GSSRealm` to 7.0.x and 8.5.x, implemented in `RealmBase` and marked as deprecated in 8.5.x. with removal in 9.0.x.
I'm neutral at this point on whether `GSSRealm` needs to extend `Realm`.
That gets everything in place one major version earlier.

Reply:
That's a good idea. I will work on that. Thank you!